File: src/graph/clonegraph.java
package graph;

import java.util.ArrayList;
import java.util.List;

public class clonegraph {

    class Node {
        public int val;
        public List<Node> neighbors;
        public Node() {
            val = 0;
            neighbors = new ArrayList<Node>();
        }
        public Node(int _val) {
            val = _val;
            neighbors = new ArrayList<Node>();
        }
        public Node(int _val, ArrayList<Node> _neighbors) {
            val = _val;
            neighbors = _neighbors;
        }
    }

//    public Node cloneGraph(Node node) {
//        if(node == null)return null;
//        var nodeMap = new HashMap<Integer, Node>();
//        var que = new LinkedList<Node>();
//        que.add(node);
//        while (!que.isEmpty()){
//            var t = que.removeFirst();
//            if (!nodeMap.containsKey(t.val))nodeMap.put(t.val, new Node(t.val));
//            if (t.neighbors != null){
//                for (Node n: t.neighbors){
//                    if (!nodeMap.containsKey(n.val)){
//                        nodeMap.put(n.val, )
//                    }
//                }
//            }
//        }
//    }

}

File: src/graph/CloneGraph.kt
package graph

import java.util.LinkedList
import kotlin.collections.HashMap

class CloneGraph {
    fun cloneGraph(node: Node?): Node? {
        if(node == null)return null
        val nodeMap = HashMap<Int?, Node?>()
        val que = LinkedList<Node>()
        que.add(node)
        while(!que.isEmpty()){
            val temp = que.poll()
            if(nodeMap[temp.`val`] == null)nodeMap[temp?.`val`] = Node(temp.`val`)
            for(i in temp.neighbors){
                if(i != null && nodeMap[i.`val`] == null){
                    val c = Node(i.`val`)
                    que.add(i)
                    nodeMap[i.`val`] = c
                }
                if (i != null){
                    nodeMap[temp.`val`]?.neighbors?.add(nodeMap[i.`val`])
                }
            }
        }
        return nodeMap[node.`val`]
    }
}

fun main(){
    val v = CloneGraph()
    val root = Node(1)
    val v2 = Node(2)
    val v3 = Node(3)
    val v4 = Node(4)
    root.neighbors.add(v2)
//    v.cloneGraph()
}